File: features/favorites/src/main/java/com/vp/favorites/FavoritesActivity.java
package com.vp.favorites;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.content.res.Configuration;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import com.vp.detail.model.MovieDetail;

import java.util.ArrayList;

import io.realm.Realm;
import io.realm.RealmResults;

public class FavoritesActivity extends AppCompatActivity implements FavoritesAdapter.OnItemClickListener {

    private RecyclerView recyclerView;
    private TextView tvEmptyMessage;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_favorite);

        recyclerView = findViewById(R.id.recyclerView);
        tvEmptyMessage = findViewById(R.id.tvEmptyMessage);

        recyclerView.setHasFixedSize(true);
        GridLayoutManager layoutManager = new GridLayoutManager(this,
                getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT ? 2 : 3);
        recyclerView.setLayoutManager(layoutManager);
    }

    /**
     * Get and display all saved favorites
     */
    private void fetchFavorites() {
        ArrayList<MovieDetail> favoriteMovies = new ArrayList<>();
        final Realm realm = Realm.getDefaultInstance();
        final RealmResults<MovieDetail> resultMovies = realm.where(MovieDetail.class).findAll();

        if (resultMovies.isEmpty()) {
            favoriteMovies.clear();
            showEmptyView();
        } else {
            favoriteMovies.addAll(realm.copyFromRealm(resultMovies));
            setFavorites(favoriteMovies);
        }

        realm.close();
    }

    /**
     * Set favorites list in adapter
     */
    private void setFavorites(ArrayList<MovieDetail> favoriteMovies) {
        recyclerView.setVisibility(View.VISIBLE);
        final FavoritesAdapter favoritesAdapter = new FavoritesAdapter();
        favoritesAdapter.setOnItemClickListener(this);
        favoritesAdapter.setItems(favoriteMovies);
        recyclerView.setAdapter(favoritesAdapter);
    }

    /**
     * Displays a message if no favorites found
     */
    private void showEmptyView() {
        recyclerView.setVisibility(View.GONE);
        tvEmptyMessage.setVisibility(View.VISIBLE);
        tvEmptyMessage.setText(getResources().getString(R.string.no_data));
    }

    @Override
    public void onItemClick(String imdbID) {
        makeDetailIntent(imdbID);
    }

    @Override
    protected void onResume() {
        fetchFavorites();
        super.onResume();
    }

    /**
     * Displays the detail activity.
     * @param imdbID the ID of the movie to display.
     */
    private void makeDetailIntent(String imdbID) {
        String deeplink = getResources().getString(R.string.detail_deeplink) + imdbID;
        Intent intent = new Intent (Intent.ACTION_VIEW);
        intent.setData(Uri.parse(deeplink));
        startActivity(intent);
    }
}

File: features/favorites/src/main/java/com/vp/favorites/FavoritesActivity.kt
package com.vp.favorites

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.res.Configuration
import android.net.Uri
import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.GridLayoutManager
import com.vp.favorites.databinding.ActivityFavoritesBinding
import com.vp.favorites.model.ListState
import com.vp.favorites.model.SearchResult
import com.vp.favorites.viewmodel.FavoritesViewModel
import dagger.android.support.DaggerAppCompatActivity
import kotlinx.android.synthetic.main.activity_favorites.*
import javax.inject.Inject

class FavoritesActivity : DaggerAppCompatActivity(), ListAdapter.OnItemClickListener {

    @Inject
    lateinit var factory: ViewModelProvider.Factory

    lateinit var sharedPreferences: SharedPreferences
    private lateinit var favoritesViewModel: FavoritesViewModel
    private lateinit var listAdapter: ListAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding: ActivityFavoritesBinding = DataBindingUtil.setContentView(this, R.layout.activity_favorites)
        favoritesViewModel = ViewModelProviders.of(this, factory).get(FavoritesViewModel::class.java)
        binding.setLifecycleOwner(this)
        supportActionBar?.title = getString(R.string.favorites)

        setupListAdapter()
        setupSwipeRefresh()

        sharedPreferences = getSharedPreferences("favorites", Context.MODE_PRIVATE)

        favoritesViewModel.favorites.observe(this, Observer {
            handleResult(listAdapter, it)
        })
    }

    override fun onStart() {
        super.onStart()
        getFavorites()
    }

    private fun getFavorites() {
        val setIds = sharedPreferences.getStringSet("movieIds", null)?.toSet()
        favoritesViewModel.fetchFavorites(setIds)
    }

    private fun setupListAdapter() {
        listAdapter = ListAdapter()
        listAdapter.setOnItemClickListener(this)
        recyclerView.adapter = listAdapter
        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = GridLayoutManager(this, if (resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT) 2 else 3)
    }

    private fun setupSwipeRefresh() {
        swipe_refresh.setOnRefreshListener {
            listAdapter.clearItems()
            getFavorites()
            swipe_refresh.isRefreshing = false
        }
    }

    override fun onItemClick(imdbID: String?) {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse("app://movies/detail?imdbID=$imdbID"))
        intent.setPackage(packageName)
        startActivity(intent)
    }

    private fun handleResult(listAdapter: ListAdapter, searchResult: SearchResult) {
        when (searchResult.listState) {
            ListState.LOADED -> {
                listAdapter.setItems(searchResult.items.toMutableList())
                showList()
            }
            ListState.IN_PROGRESS -> {
                showProgressBar()
            }
            ListState.NO_DATA -> {
                showNoFavoritesMessage()
            }
            else -> {
                showError()
            }
        }
    }

    private fun showProgressBar() {
        viewAnimator.displayedChild = viewAnimator.indexOfChild(progressBar)
    }

    private fun showList() {
        viewAnimator.displayedChild = viewAnimator.indexOfChild(recyclerView)
    }

    private fun showError() {
        viewAnimator.displayedChild = viewAnimator.indexOfChild(errorText)
    }

    private fun showNoFavoritesMessage() {
        viewAnimator.displayedChild = viewAnimator.indexOfChild(noFavoritesText)
    }
}